sycl/include/sycl/types.hpp

```diff
@@ -789,18 +788,40 @@ template <typename Type, int NumElements> class vec {
     using T = vec_data_t<DataT>;
     using R = vec_data_t<convertT>;
     vec<convertT, NumElements> Result;
+
+    // Whole vector conversion can only be done, if:
+    constexpr bool canUseNativeVectorConvert =
+#if defined(__INTEL_PREVIEW_BREAKING_CHANGES)
+    // - we are in preview breaking changes mode
+#ifdef __NVPTX__
+        // - we are not on CUDA, see intel/llvm#11840
+        false &&
+#endif
+        // - both vectors are represented using native vector types;
+        NativeVec && vec<convertT, NumElements>::NativeVec &&
+        // - it is not a signed to unsigned (or vice versa) conversion
+        //   see comments within 'convertImpl' for more details;
+        !detail::is_sint_to_from_uint<T, R>::value &&
+        // - destination type is not bool. bool is stored as integer under the
+        //   hood and therefore conversion to bool looks like conversion between
+        //   two integer types. Since bit pattern for true and false is not
+        //   defined, there is no guarantee that integer conversion yields
+        //   right results here;
+        !std::is_same_v<convertT, bool>;
+#else
+        // vec<T, N>::NativeVec is not available
+        false;
+#endif // __INTEL_PREVIEW_BREAKING_CHANGES
+
+    if constexpr (canUseNativeVectorConvert) {
 #if defined(__INTEL_PREVIEW_BREAKING_CHANGES)
```

Comment:
If I understand correctly, `canUseNativeVectorConvert` can only be true when `__INTEL_PREVIEW_BREAKING_CHANGES` is defined, so why do we need to check this here again?

Reply:
`vec<T, N>::VectorDataType` is only available in preview breaking changes mode, unfortunately. I have a separate refactoring patch which makes it available everywhere here: #11879 and it should allow to remove that `#ifdef`.

At the same time, I remember that @steffenlarsen said that he will do some changes around `VectorDataType` and I see conflict now that his patch is merged. Let me do merge with `intel/llvm` first to see if we can improve the situation here already

Reply:
I've just pushed a merge commit which unified those changes with ones done in #11883 and it reduced amount of macro-branches we have